pkg/git/git_error_parser.go

```diff
@@ -0,0 +1,288 @@
+package git
+
+import (
+	"bufio"
+	"errors"
+	"regexp"
+	"strings"
+)
+
+type (
+	// ErrorClass classifies git stdout error messages in broader categories
+	ErrorClass int
+	// Prefix is part of an error message output and can be used to determine which participant in the git protocol
+	// send parts of the error message
+	Prefix     int
+)
+
+const (
+	unknownPrefix Prefix = iota
+	remotePrefix
+	fatalPrefix
+	errorPrefix
+)
+
+const (
+	// Unknown is the class of choice if no other class fits
+	Unknown ErrorClass = iota
+	// AuthInvalidUserOrPass expresses that basic authentication is not possible
+	AuthInvalidUserOrPass
+	// AuthInvalidKey expresses that ssh authentication is not possible
+	AuthInvalidKey
+	// BranchNotFound expresses that a remote branch does not exist.
+	BranchNotFound
+	// RepositoryNotFound expresses that the remote target for the git operation does not exist. It triggers when an
+	// error message is enough to determine that the remote target does not exist and is mostly derived from the
+	// Git server's messages e.g. GitLab or GitHub.
+	RepositoryNotFound
+	// AuthPrompted is caused when a repo is not found, is private and authentication is insufficient
+	AuthPrompted
+)
+
+type rawToken struct {
+	raw string
+}
+
+type prefixToken struct {
+	scope Prefix
+	rawToken
+}
+
+type errorClassToken struct {
+	class ErrorClass
+	rawToken
+}
+
+type errorToken struct {
+	prefixToken prefixToken
+	classToken  errorClassToken
+}
+
+// ErrorResult is a representation of a runtime error of a git operation that presents a reason and a message
+type ErrorResult struct {
+	Message string
+	Reason  ErrorClass
+}
+
+func (rawToken rawToken) String() string {
+	return rawToken.raw
+}
+
+func (class ErrorClass) String() string {
+	switch class {
+	case AuthInvalidUserOrPass:
+		return "git-auth-basic"
+	case AuthInvalidKey:
+		return "git-auth-ssh"
+	case BranchNotFound:
+		return "git-remote-revision"
+	case RepositoryNotFound:
+		return "git-remote-repository"
+	case AuthPrompted:
+		return "git-remote-private"
+	}
+
+	return "git-error"
+}
+
+func (class ErrorClass) toMessage() string {
+	switch class {
+	case RepositoryNotFound:
+		return "The source repository does not exist or you have insufficient rights."
+	case AuthInvalidUserOrPass:
+		return "Basic authentication has failed. Check your username or password. Note: Github requires a personal access token instead of your regular password."
+	case AuthPrompted:
+		return "The source repository does not exist or you have insufficient right. Provide authentication for more details."
+	case BranchNotFound:
+		return "The remote branch does not exist. Check your revision argument"
```

Comment:
As suggested in the ship, this might also be a tag name, then more changes are necessary.

```suggestion
		return "The remote branch does not exist. Check your revision argument."
```